runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt

```diff
@@ -19,30 +19,30 @@ import aws.smithy.kotlin.runtime.text.encoding.encodeBase64String
  * See:
  *   - https://awslabs.github.io/smithy/1.0/spec/core/behavior-traits.html#httpchecksumrequired-trait
  *   - https://datatracker.ietf.org/doc/html/rfc1864.html
+ * @param block An optional function which parses the input [I] to determine if the `Content-MD5` header should be set.
+ * If not provided, the default behavior will set the header.
  */
 @InternalApi
 public class Md5ChecksumInterceptor<I>(
     private val block: ((input: I) -> Boolean)? = null,
 ) : HttpInterceptor {
+    private var cachedChecksum: String? = null
```

Comment:
I like the idea, but I think it would be an API break. Users making custom `Md5ChecksumInterceptor` / `FlexibleChecksumsRequestIntereceptor` would need to implement these new methods (`calculateChecksum` / `applyChecksum`).

Reply:
Good point, I didn't fully think this through... I can try the refactor and see how it looks

Reply:
FYI: I did the refactor and the .api file appears "broken" because the implementation changed from `HttpInterceptor` to `AbstractChecksumInterceptor` but under the hood it's still valid